pkg/gadgets/trace/dns/tracer/bpf/dns.c

```diff
@@ -114,159 +102,136 @@ struct {
+				break;
+			case NEXTHDR_HOP:
+			case NEXTHDR_ROUTING:
+			case NEXTHDR_DEST:
+				// See ipv6_optlen()
+				l4_off += 8 * (load_byte(skb, l4_off + 1) + 1);
+				break;
+			case NEXTHDR_NONE:
+				// Nothing more in the packet. Not even TCP or UDP.
+				return 0;
+			default:
+				// Unknown header
+				return 0;
+			}
+			proto = nextproto;
 		}
-		rroffset += sizeof(struct dnsrr) + rdlength;
-	}
-	return index;
-}
+		break;
 
-static __always_inline int output_dns_event(struct __sk_buff *skb,
-					    union dnsflags flags,
-					    __u32 name_len, __u16 ancount)
-{
-	__u32 zero = 0;
-	struct event_t *event = bpf_map_lookup_elem(&tmp_event, &zero);
-	if (!event)
+	default:
 		return 0;
-
-	__builtin_memset(event, 0, sizeof(*event));
-
-	event->netns = skb->cb[0]; // cb[0] initialized by dispatcher.bpf.c
-	event->timestamp = bpf_ktime_get_boot_ns();
-	event->id = load_half(skb, DNS_OFF + offsetof(struct dnshdr, id));
-	event->af = AF_INET;
-	event->daddr_v4 =
-		load_word(skb, ETH_HLEN + offsetof(struct iphdr, daddr));
-	event->saddr_v4 =
-		load_word(skb, ETH_HLEN + offsetof(struct iphdr, saddr));
-	// load_word converts from network to host endianness. Convert back to
-	// network endianness because inet_ntop() requires it.
-	event->daddr_v4 = bpf_htonl(event->daddr_v4);
-	event->saddr_v4 = bpf_htonl(event->saddr_v4);
-
-	// Check network protocol.
-	// This only works with IPv4.
-	// For IPv6, gadget_socket_lookup() in sockets-map.h
-	// provides an example how to parse ip/ports on IPv6.
-	event->proto =
-		load_byte(skb, ETH_HLEN + offsetof(struct iphdr, protocol));
-	if (event->proto == IPPROTO_TCP) {
-		event->sport =
-			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
-					       offsetof(struct tcphdr, source));
-		event->dport =
-			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
-					       offsetof(struct tcphdr, dest));
-	} else if (event->proto == IPPROTO_UDP) {
-		event->sport =
-			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
-					       offsetof(struct udphdr, source));
-		event->dport =
-			load_half(skb, ETH_HLEN + sizeof(struct iphdr) +
-					       offsetof(struct udphdr, dest));
 	}
 
-	event->qr = flags.qr;
-
-	if (flags.qr == 1) {
-		// Response code set only for replies.
-		event->rcode = flags.rcode;
+	switch (proto) {
+	case IPPROTO_UDP:
+		event.sport = load_half(skb, l4_off + offsetof(struct udphdr,
+							       source));
+		event.dport =
+			load_half(skb, l4_off + offsetof(struct udphdr, dest));
+		dns_off = l4_off + sizeof(struct udphdr);
+		break;
+	// TODO: support TCP
+	default:
+		return 0;
 	}
 
-	bpf_skb_load_bytes(skb, DNS_OFF + sizeof(struct dnshdr), event->name,
-			   name_len);
+	event.netns = skb->cb[0]; // cb[0] initialized by dispatcher.bpf.c
+	event.timestamp = bpf_ktime_get_boot_ns();
+	event.proto = proto;
+	event.dns_off = dns_off;
 
-	event->pkt_type = skb->pkt_type;
+	if (!is_dns_port(event.sport) && !is_dns_port(event.dport))
+		return 0;
```

Comment:
Makes sense, updated the code. The only think I didn't change was the allocation of the event structure, it's being allocated on the stack, so I don't think "allocating" it later has a visible effect on performance.

Reply:
I _think_ you can at least prevent the zeroing operation by using `struct event_t event;` and zero it manually later on, rather than using `struct event_t event = {};`, which implies it, right?